FIX #15016 File name with two spaces throws ErrorFileDoesNotExist

Ahuahuachi opened the pull request:
GETPOST applied to file names should not need an alphanohtml check

Ahuahuachi committed:
FIX #15016 File name with two spaces throws ErrorFileDoesNotExist

eldy commented:
Allowing any char leads to a lot of security issues. Also with v13 alpha = alphahtml.
So i think we should fix differently. I tested with develop branch and i do not experience any trouble when uploading a file as a join file of a document. Did you find a way to reproduce the problem ?

Ahuahuachi commented:
Yes I reproduced the problem, it is present from branch 10.0 through 12.0.
It is specifically caused by the `dol_string_nohtmltag` function which is stripping double spaces in the following lines https://github.com/Dolibarr/dolibarr/blob/c918acaadc3571d2186ec489a0de5ff5a6cda140/htdocs/core/lib/functions.lib.php#L5648-L5652
This function is called by `GETPOST` when check type is `alphanohtml`
When the file name has double spaces the reference to it gets replaced by a single space, throwing the mentioned error when trying to preview it.
Another way around would be to make doble space stripping optional?

eldy commented:
Can you describe how you reproduce. I failed on develop branch.

Ahuahuachi commented:
The following are the tests that I made.

1. I renamed a file that was already uploaded/created from an object (in my case a customer invoice) via a file explorer, adding 2 spaces in between. In this test the problem was successfully reproduced.

2. I uploaded a file with two consecutive spaces in its name unchecking the option of renaming. In this scenario the problem did not appeared, because when the file gets uploaded it gets renamed converting the double spaces into one.

3. I renamed a file via the "Linked files" tab adding double spaces in between the name. In this case the problem was indeed reproduced.

eldy commented:
Thanks. I fixed case 3 for consistency with 2.
For security purpose we must sanitize to remove html content. This has also effect to replace duplicate spaces into 1 space. But this is not a so bad effect, if we have consistency.
I however introduced a new parameter removedoublespaces into the dol_string_nohtmltag() method for future use. 623cb54480c68ad213ae6a128a03c4c278bf5441

About case 1, do you mean you rename file on disk directly into "documents" directory ? This is les annoying than case 3 but we may solve this with the new option introduced.
